### PR TITLE
Invalid data

### PR DESCRIPTION
**Pull request by katarzynajaskolska**
To check if files in Permissions repo contain valid information:

```ruby
DataGuru::Client.new.invalid_values
```

You will get a hash with names of invalid files and values which are missing or have invalid datatype:

```ruby
# {
#   "john.doe.yml" => {
#     "missing_attributes" => [:emails, :external],
#     "invalid_data_type_attributes" => [:github]
#   }
# }
```

To check invalid values of particular collection:

```ruby
DataGuru::Client.new.invalid_values("users")
```

To check particular value in particular collection:

```ruby
DataGuru::Client.new.invalid_values("users", "name")
# array containing names of files from "users" collection that have invalid "name" value
```

## TODO:

- [x] specs
- [x] validation for data types (e.g. check if values are strings if strings are required)
- [x] check if it is possible to limit the number of calls to api
- [x] update Readme
- [x] handle passing non-existing collection name and value name + specs

### CODE REVIEW

File: lib/data_guru/model.rb [hunk 1]
```diff
@@ -41,7 +54,7 @@ def set_attr_value(attribute, value, default_value)
 
     def present_required_attributes
       required_attributes.map do |name|
-        !attributes[name].nil? ? name : nil
+        attributes[name].present? ? name : nil
```
Comment (lubieniebieski) [inline lib/data_guru/model.rb]: much better

File: README.md [hunk 2]
```diff
@@ -87,18 +87,62 @@ member = data.members.all.first
 # to get list of member attributes
 member.attributes
 
-# to get list of permitted and required attributes (see Permisions section below)
+# to get list of permitted and required attributes (see Permissions section below)
 member.permitted_attributes
 member.required_attributes
 
-# check if member has all required attributes set
+# check if member has all required attributes set and they have valid datatype
```
Comment (zatapathique) [inline README.md]: maybe `check if member has all required attributes set with valid datatypes`

File: lib/data_guru/client.rb [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
-    require 'active_support/core_ext/string'
+    require "httparty"
+    require "active_support/core_ext/string"
 
     def initialize(api_url: nil, access_token: nil)
-      if api_url.present? && access_token.present?
-        DataGuru.config = Configuration.new(api_url: api_url,
-                                            access_token: access_token)
-      end
+      return unless api_url.present? && access_token.present?
```
Comment (zatapathique) [inline lib/data_guru/client.rb]: `Unless` isn't really a strong point of ruby :) If you can describe something with `if` it is usually a good idea to do so - `if's` are easier to parse mentally, every negation just adds to the process thus I think that you could rewrite this as:
`return if api_url.blank? || access_token.blank?`

File: lib/data_guru/invalid_values.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      h.map { |k, v| k if v.to_s.include?(attribute.to_s) }.reject(&:nil?)
+    end
+
+    def value_exists?
+      data.send(collection_name).config_data.keys.include? attribute.to_s
+    end
+
+    def value_not_exists_error
+      "Value '#{attribute}' does not exist for '#{collection_name}' collection"
+    end
```
Comment (zatapathique) [inline lib/data_guru/invalid_values.rb]: If you find yourself adding lots of custom error messages you can always consider adding module with custom error classes:
```
module Errors
  class MissingValue < StandadrError
     de message; 'some message'; end
  end
end
```
This approach separates application logic with error presentation.

File: lib/data_guru/model.rb [hunk 5]
```diff
@@ -19,13 +20,25 @@ def attributes
...
-      present_required_attributes == required_attributes
+      missing_attributes.empty? && invalid_data_type_attributes.empty?
     end
 
     def missing_attributes
       required_attributes - present_required_attributes
     end
 
+    def invalid_data_type_attributes
+      data_types.map do |k, v|
```
Comment (zatapathique) [inline lib/data_guru/model.rb]: You could do `key = k.to_sym` or even better add some custom, informative name - `name = k.to_sym`.

File: lib/data_guru/model.rb [hunk 6]
```diff
@@ -19,13 +20,25 @@ def attributes
...
+        next if attributes[k.to_sym].nil?
+        if v == "boolean"
+          k.to_sym unless !!attributes[k.to_sym] == attributes[k.to_sym]
```
Comment (zatapathique) [inline lib/data_guru/model.rb]: This is quite hard to parse - if I got it right this condition checks if attribute is either `true` or `false`directly and not a string `'false'` or any other class. You could rewrite it for clarity:
`name if [true, false].exclude?(attribute)` or `name unless attribute.in? [true, false]`

File: lib/data_guru/model.rb [hunk 7]
```diff
@@ -41,13 +54,13 @@ def set_attr_value(attribute, value, default_value)
...
-        !attributes[name].nil? ? name : nil
+        attributes[name].present? ? name : nil
       end
     end
 
     def all_attributes
       attrs = permitted_attributes
-        .map{ |name| [name, public_send(name)] } << [:id, id]
+              .map { |name| [name, public_send(name)] } << [:id, id]
       Hash[attrs]
```
Comment (zatapathique) [inline lib/data_guru/model.rb]: You could just directly work on hash:
`permitted_attributes.each_with_object({ id: id }) do |name, hash|`

File: lib/data_guru/model_configuration.rb [hunk 8]
```diff
@@ -1,32 +1,21 @@
...
-      @model_name = model_name
-    end
-
-    def attributes(config_data)
-      config_data[model_name]
     end
 
     def permitted_attributes(config_data)
-      attributes(config_data).map{ |k, _v| k.to_sym }
+      config_data.map { |k, _v| k.to_sym }
```
Comment (zatapathique) [inline lib/data_guru/model_configuration.rb]: `config_data .symbolize_keys.keys` :)

File: lib/data_guru/model_configuration.rb [hunk 9]
```diff
@@ -1,32 +1,21 @@
...
     end
 
     def required_attributes(config_data)
-      attributes(config_data).map{ |k, v| v["required"] ? k.to_sym : nil }.reject(&:nil?)
+      config_data.map { |k, v| v["required"] ? k.to_sym : nil }.reject(&:nil?)
```
Comment (zatapathique) [inline lib/data_guru/model_configuration.rb]: you could also do it like this:
`config_data.select{ |_k,v| v['required'] }.symbolize_keys.keys`

File: lib/data_guru/model_configuration.rb [hunk 10]
```diff
@@ -1,32 +1,21 @@
...
-    private
-
-    def default_model_name(model)
-      model.name.split('::').last
-        .gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2')
-        .gsub(/([a-z\d])([A-Z])/,'\1_\2')
-        .tr("-", "_")
-        .downcase
+    def data_types(config_data)
+      config_data.each_with_object({}) { |(k, v), h| h[k] = v["value_type"] }
```
Comment (zatapathique) [inline lib/data_guru/model_configuration.rb]: Seems like this class heavily relies on `config_data`, maybe you should just move it to the initializer so the class can store it and use it within the methods without the need to supply it to every single method? It would also solve the problem of symbolizing the keys - you symbolize key everytime you use config_data - its cumbersome and prone to errors. You could just symbolize the keys in initializer:
'@config_data = config_data.symbolize_keys`
It is doable? :)

File: lib/data_guru/validations/model_configuration.rb [hunk 11]
```diff
@@ -36,41 +36,41 @@ def build_errors
...
+      keys.include?("config")
     end
 
     def model_file_valid?
       return false unless config_valid?
       valid_files = model_names
-        .map{ |name| config.include?(name) }
-        .select{ |name| name }
+                    .map { |name| config.include?(name) }
+                    .select { |name| name }
```
Comment (zatapathique) [inline lib/data_guru/validations/model_configuration.rb]: If you just want the number of the names included in the config you can do:
`model_names .count { |name| config.include?(name) }`

File: lib/data_guru/validations/model_configuration.rb [hunk 12]
```diff
@@ -36,41 +36,41 @@ def build_errors
...
       valid_files.size == model_names.size
     end
 
     def model_attributes_valid?
       return false unless model_file_valid?
       valid_models = model_names
-        .map{ |name| config[name] == false || config[name].nil? }
-        .select{ |name| !name }
+                     .map { |name| config[name] == false || config[name].nil? }
+                     .select(&:!)
```
Comment (zatapathique) [inline lib/data_guru/validations/model_configuration.rb]: This is hard to read/parse - you can replace it with `model_names.count { |name| config[name].present? }`
Double check this please because there may be some edge cases with this `false / nil` checks, ie. if previously `config[name]` was equal `''` then it would be classified as valid, but when you check presence on an empty string it returns false.

File: lib/data_guru/validations/model_configuration.rb [hunk 13]
```diff
@@ -36,41 +36,41 @@ def build_errors
...
 
     def model_attribute_content_valid?
-      attributes = present_models.map{ |model| config[model] }
-      invalid_attrs = attributes.flat_map{ |model|
-          model.values.map{ |attributes| attributes.keys.sort == ATTRIBUTE_TYPES.sort }
-        }.select{ |a| !a }
+      attributes = present_models.map { |model| config[model] }
+      invalid_attrs = attributes.flat_map do |model|
+        model.values.map { |attrs| attrs.keys.sort == ATTRIBUTE_TYPES.sort }
+      end.select(&:!)
```
Comment (zatapathique) [inline lib/data_guru/validations/model_configuration.rb]: `select(&:!)` is very hard to parse, I think you should replace it with some sort of `count { }` and some smart conditionals.

File: lib/data_guru/validations/model_configuration.rb [hunk 14]
```diff
@@ -119,21 +121,21 @@ def present_models
...
         wrong_attrs = config[model].keys & RESERVED_WORDS
-        [ model, wrong_attrs ]
+        [model, wrong_attrs]
       end
       Hash[wrong_attributes]
     end
 
     def missing_model_attributes
-      all = Hash[present_models.map{ |model| [model, config[model]] }]
+      all = Hash[present_models.map { |model| [model, config[model]] }]
```
Comment (zatapathique) [inline lib/data_guru/validations/model_configuration.rb]: You can use `each_with_object` instead of Array -> Hash conversion

File: lib/data_guru/validations/model_configuration.rb [hunk 15]
```diff
@@ -119,21 +121,21 @@ def present_models
...
-          without_empty = attrs
-            .map{ |name, types| types.empty? ? [] : [name, types] }
-            .reject{ |array| array.empty? }
-          [ model, Hash[without_empty] ]
-        end
-      Hash[missing.reject{ |k, v| v.empty? }]
+        attrs = Hash[attributes.map { |name, types| [name, ATTRIBUTE_TYPES - types.keys] }]
+        without_empty = attrs
+                        .map { |name, types| types.empty? ? [] : [name, types] }
+                        .reject(&:empty?)
```
Comment (zatapathique) [inline lib/data_guru/validations/model_configuration.rb]: `reject` also works on hashes so you don't have to do any maping, you can call reject/select straight on a hash.

File: lib/data_guru/validations/model_configuration.rb [hunk 16]
```diff
@@ -119,21 +121,21 @@ def present_models
...
+        [model, Hash[without_empty]]
+      end
+      Hash[missing.reject { |_k, v| v.empty? }]
```
Comment (zatapathique) [inline lib/data_guru/validations/model_configuration.rb]: Same here, I think that this whole method will be more readable if you rewrite it with `each_with_object` and `reject/select` using only hashes.

File: lib/data_guru/invalid_values.rb [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        invalid_files << file unless file.valid?
+      end
+    end
+
+    def collection_names
+      collection_name.present? ? [collection_name] : data.send(:collection_names) - ["config"]
+    end
+
+    def files_with_invalid_attribute(attribute, h)
+      fail DataGuru::MissingValue.new(attribute, collection_name).message unless value_exists?
```
Comment (zatapathique) [inline lib/data_guru/invalid_values.rb]: You don't have to call `.message` here, basically `fail` takes either a string or an `Error` instance. If given the later it will call the method `message` itself and automatically add error class to the error message ->
ie. 'DataGuru::MissingValue: values A and B are required.'. If you supply a string it will just print a string.

**Comment by zatapathique** [comment]
Review done, everything is ok :)
